player/lua.c

```diff
@@ -1194,17 +1167,64 @@ static const struct fn_entry utils_fns[] = {
     FN_ENTRY(split_path),
     FN_ENTRY(join_path),
     FN_ENTRY(getpid),
-    FN_ENTRY(parse_json),
-    FN_ENTRY(format_json),
+    AF_ENTRY(parse_json),
+    AF_ENTRY(format_json),
     {0}
 };
 
+/* returns the talloc ctx which script_autofree_trampoline sets */
+static void *mp_lua_PITA(lua_State *L)
+{
+    void **ctx = (void**)lua_touserdata(L, lua_upvalueindex(1));
+    if (!ctx) {
+        lua_pushstring(L, "missing autofree ctx");
+        lua_error(L);
+    }
+    return *ctx;
+}
+
+static int script_autofree_trampoline(lua_State *L)
+{
+    lua_CFunction target = lua_touserdata(L, lua_upvalueindex(1));
+    if (!target) {
+        lua_pushstring(L, "missing autofree target");
+        lua_error(L);
+    }
+
+    int nargs = lua_gettop(L);
+    void *ctx;
+
+    lua_pushlightuserdata(L, &ctx);
+    lua_pushcclosure(L, target, 1);
+    for (int i = 1; i <= nargs; i++)
+        lua_pushvalue(L, i);
```

Comment:
You may be able to use a Lua API function that inserts the stack items to before the arguments or so. Generally, redirection a call (sort of like a tail call) should be possible without having to push all arguments on the stack again. But not sure.

Reply:
I'll look into it, and it would definitely be nicer, at the very least for needing less C code.